src/Platform/Microsoft.Testing.Platform/Hosts/CommonTestHost.cs

```diff
@@ -26,25 +33,40 @@ public async Task<int> RunAsync()
         int exitCode;
         try
         {
-            if (RunTestApplicationLifeCycleCallbacks)
+            bool isDotnetTestHandshakeSuccessful = default;
+            ITestApplicationModuleInfo testApplicationModuleInfo = serviceProvider.GetTestApplicationModuleInfo();
+
+            if (dotnetTestPipeClient is not null)
+            {
+                isDotnetTestHandshakeSuccessful = await IsDotnetTestHandshakeSuccessfulAsync(dotnetTestPipeClient, testApplicationModuleInfo, testApplicationCancellationToken);
+            }
+
+            if (isDotnetTestHandshakeSuccessful)
             {
-                // Get the test application lifecycle callbacks to be able to call the before run
-                foreach (ITestApplicationLifecycleCallbacks testApplicationLifecycleCallbacks in ServiceProvider.GetServicesInternal<ITestApplicationLifecycleCallbacks>())
+                if (RunTestApplicationLifeCycleCallbacks)
                 {
-                    await testApplicationLifecycleCallbacks.BeforeRunAsync(testApplicationCancellationToken);
+                    // Get the test application lifecycle callbacks to be able to call the before run
+                    foreach (ITestApplicationLifecycleCallbacks testApplicationLifecycleCallbacks in ServiceProvider.GetServicesInternal<ITestApplicationLifecycleCallbacks>())
+                    {
+                        await testApplicationLifecycleCallbacks.BeforeRunAsync(testApplicationCancellationToken);
+                    }
                 }
-            }
 
-            exitCode = await InternalRunAsync();
+                exitCode = await InternalRunAsync();
 
-            if (RunTestApplicationLifeCycleCallbacks)
-            {
-                foreach (ITestApplicationLifecycleCallbacks testApplicationLifecycleCallbacks in ServiceProvider.GetServicesInternal<ITestApplicationLifecycleCallbacks>())
+                if (RunTestApplicationLifeCycleCallbacks)
                 {
-                    await testApplicationLifecycleCallbacks.AfterRunAsync(exitCode, testApplicationCancellationToken);
-                    await DisposeHelper.DisposeAsync(testApplicationLifecycleCallbacks);
+                    foreach (ITestApplicationLifecycleCallbacks testApplicationLifecycleCallbacks in ServiceProvider.GetServicesInternal<ITestApplicationLifecycleCallbacks>())
+                    {
+                        await testApplicationLifecycleCallbacks.AfterRunAsync(exitCode, testApplicationCancellationToken);
+                        await DisposeHelper.DisposeAsync(testApplicationLifecycleCallbacks);
+                    }
                 }
             }
+            else
+            {
+                exitCode = ExitCodes.IncompatibleProtocolVersion;
```

Comment:
You could have a first return directly at the start of the method when the pipe is null, this would simplify the rest of this method